src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,811 @@
+    if (const auto* op = expr.as_derived<CanonicalExprNode>()) {
+      return op->Normalize();
+    } else {
+      return expr;
+    }
+  }
+  /*!
+   * \brief Create a SplitExpr from expr.
+   * \param expr The input expr.
+   * \return The transformed SplitExpr.
+   */
+  SplitExpr ToSplitExpr(Expr expr) {
+    if (const auto* op = expr.as<SplitExprNode>()) {
+      return GetRef<SplitExpr>(op);
+    }
+    if (const auto* op = expr.as_derived<CanonicalExprNode>()) {
+      expr = op->Normalize();
+    }
+    NodePtr<SplitExprNode> n = make_node<SplitExprNode>();
+    n->type = expr.type();
+    n->index = std::move(expr);
+    return SplitExpr(n);
+  }
+  /*!
+   * \brief Create a SumExpr from expr.
+   * \param expr The input expr.
+   * \return The transformed SumExpr.
+   */
+  SumExpr ToSumExpr(Expr expr) {
+    if (const auto* op = expr.as<SumExprNode>()) {
+      return GetRef<SumExpr>(op);
+    }
+    NodePtr<SumExprNode> n = make_node<SumExprNode>();
+    n->type = expr.type();
+    if (const auto* op = expr.as<IntImm>()) {
+      n->base = op->value;
+      return SumExpr(n);
+    } else {
+      n->args.emplace_back(ToSplitExpr(expr));
+      return SumExpr(n);
+    }
+  }
+  // Simplify the combiner used in reduce.
+  Expr SimplifyReduceCombiner(const Reduce* op);
+};
+
+Expr CanonicalSimplifier::Impl::
+Mutate_(const Add* op, const Expr& self) {
+  if (!IsIndexType(op->type)) {
+    return Rewriter::Mutate_(op, self);
+  }
+  // normalize
+  Expr a = this->CanonicalMutate(op->a);
+  Expr b = this->CanonicalMutate(op->b);
+
+  // const folding
+  Expr const_res = TryConstFold<Add>(a, b);
+  if (const_res.defined()) return const_res;
+
+  // canonical form simplification.
+  SumExpr ret = ToSumExpr(std::move(a));
+
+  if (const auto* op = b.as<IntImm>()) {
+    ret.CopyOnWrite()->AddToSelf(op->value);
+  } else if (const auto* op = b.as<SumExprNode>()) {
+    ret.CopyOnWrite()->AddToSelf(GetRef<SumExpr>(op), 1);
+  } else {
+    ret.CopyOnWrite()->AddToSelf(ToSplitExpr(b), 1);
+  }
+  return ret;
+}
+
+Expr CanonicalSimplifier::Impl::
+Mutate_(const Sub* op, const Expr& self) {
+  if (!IsIndexType(op->type)) {
+    return Rewriter::Mutate_(op, self);
+  }
+  // normalize
+  Expr a = this->CanonicalMutate(op->a);
+  Expr b = this->CanonicalMutate(op->b);
+
+  // const folding
+  Expr const_res = TryConstFold<Sub>(a, b);
+  if (const_res.defined()) return const_res;
+
+  // canonical form simplification.
+  SumExpr ret = ToSumExpr(std::move(a));
+
+  if (const auto* op = b.as<IntImm>()) {
+    ret.CopyOnWrite()->AddToSelf(-op->value);
+  } else if (const auto* op = b.as<SumExprNode>()) {
+    ret.CopyOnWrite()->AddToSelf(GetRef<SumExpr>(op), -1);
+  } else {
+    ret.CopyOnWrite()->AddToSelf(ToSplitExpr(b), -1);
+  }
+  return ret;
+}
+
+
+Expr CanonicalSimplifier::Impl::
+Mutate_(const Mul* op, const Expr& self) {
+  if (!IsIndexType(op->type)) {
+    return Rewriter::Mutate_(op, self);
+  }
+  // normalize
```

Comment:
Probably we may build a function to reduce the duplicate code in above three functions